### PR TITLE
S2589: Add reproducer for #8045

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Related to #8045

### CODE REVIEW

**Comment by Tim-Pohlmann** [comment]
I would move this test to [`CollectionConstraintTests`](https://github.com/SonarSource/sonar-dotnet/blob/176773cc9d6846edb60834a29187575d2cc53344/analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/CollectionConstraintTests.cs).

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/ConditionEvaluatesToConstantTest.cs [hunk 1]
```diff
@@ -60,6 +60,15 @@ public class ConditionEvaluatesToConstantTest
...
             .Verify();
 
+    [TestMethod]
+    public void ConditionEvaluatesToConstant_Roslyn_CS_WithS4158() =>
+        new VerifierBuilder()
+        .AddAnalyzer(() => new CS.SymbolicExecutionRunner(AnalyzerConfiguration.AlwaysEnabled))
+        .WithBasePath(@"SymbolicExecution\Roslyn")
+        .WithOnlyDiagnostics(ChecksCS.ConditionEvaluatesToConstant.S2583, ChecksCS.ConditionEvaluatesToConstant.S2589, ChecksCS.EmptyCollectionsShouldNotBeEnumerated.S4158)
+        .AddPaths("ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs")
+        .Verify();
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/ConditionEvaluatesToConstantTest.cs:70]:
```suggestion
    public void ConditionEvaluatesToConstant_Roslyn_CS_WithS4158() =>
        roslynCS.AddPaths("ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs")
        .WithOnlyDiagnostics(ChecksCS.ConditionEvaluatesToConstant.S2583, ChecksCS.ConditionEvaluatesToConstant.S2589, ChecksCS.EmptyCollectionsShouldNotBeEnumerated.S4158)
        .Verify();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
+
+namespace Tests.Diagnostics
+{
+    class Tests
+    {
+
+        private int Length { get; }
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs:8]:
```suggestion
    {
        private int Length { get; }
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
+
+namespace Tests.Diagnostics
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs:3]: Change it to file-scoped.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs (hunk 2 above)
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs:8]:
```suggestion
        int Length { get; }
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+{
+    class Tests
+    {
+
+        private int Length { get; }
+
+        void Test(List<string> foo)
+        {
+            bool result = foo.Contains("foo");
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs:12]: Add a comment to explain that this is needed to trigger the rule via `ShouldExecute`.

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
As mentioned above, I would merge this with the other test file.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs (hunk 2 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs:8]: obsolete

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs (hunk 3 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs:3]: obsolete

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs (hunk 2 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.EmptyCollectionsShouldNotBeEnumerated.cs:8]: obsolete

**Pull request merged**